### ISSUE TITLE
Clarification on online-softmax correctness and baseline comparability

### ISSUE THREAD

**Issue by Qi-Zhan** [kaist-ina/Trinity-AE#2]
Dear Trinity authors,

I am studying Trinity and trying to reproduce the attention optimization in
Section 5.3 and the reported baseline comparisons. I have several questions
about the numerical implementation and evaluation methodology.

In Section 5.3 and Appendix C, the paper describes FlashAttention-style online
softmax with a running maximum, running sum, and rescaling of the previous
output accumulator. However, the generated kernel in the artifact, for example

backend/results/vanilla/vanilla_llama* appears to implement:

    C_exp = tl.exp(tl.dot(Q, K).to(tl.float32))
    C_sum += tl.sum(C_exp, axis=2)
    O = tl.dot(C_exp, V) + O
    ...
    O = O / C_sum[:, :, None]

I do not see a row-wise maximum, subtraction of the running maximum, or
rescaling of the previous sum and output accumulator.

This distinction also seems important for the FlashInfer comparison. In
backend/baselines.py, the FlashInfer baseline invokes:

    flashinfer.single_prefill_with_kv_cache(
        ...,
        sm_scale=1.0
    )

this API dispatches to a FlashAttention-style FA2/FA3
backend and performs numerically stable fused attention. Therefore, unless the
running-maximum operations are inserted elsewhere, the Trinity and FlashInfer
kernels do not appear to perform comparable finite-precision algorithms, which is unfaire comparison.

Could you clarify the following?

1. Were the reported Trinity results obtained from a kernel containing the
   full online-softmax update, including running maximum and accumulator
   rescaling?

2. What correctness checks were performed against a stable reference? In
   particular, are there tests using realistic model weights or larger logits
   that check relative error, Inf, and NaN behavior?

3. Would the performance comparison change if Trinity were required to
   implement the same numerically stable softmax semantics as FlashInfer?

Any pointers to the relevant commits, generated kernels, or validation scripts
would be greatly appreciated.

**Comment by jhpark676** [comment]
1. You are correct. The reported results were obtained from a kernel that does not perform rescaling via a running maximum.

2. However, we verified that at the relatively short context length used in our experiments (1024), this does not compromise numerical stability.
Across a variety of random inputs, we confirmed that the outputs agree with a PyTorch reference implementation within an acceptable tolerance.

3. We also confirmed that adding maximum-based rescaling does not significantly change latency. Since the dominant bottleneck in latency is matmul computation, the presence or absence of maximum rescaling makes little difference.
We therefore believe the comparison against FlashInfer remains meaningful, though we agree the paper should have stated this distinction explicitly.

4. Nevertheless, we recognize that as the context length grows larger, ensuring numerical stability through maximum rescaling becomes an important issue.
To address this, we added additional rewrite rules related to maximum rescaling, and we have resolved the problem this way. (not yet publicly released)

We are continuously upgrading Trinity, and we plan to release an improved version that resolves the issues you raised in the near future. Thank you once again for your interest in our work.
